### PR TITLE
Support fetching from URLs

### PR DESCRIPTION
**Pull request by imjasonh**
This adds:

- `sget <URL>` to fetch URLs iff a configured trusted identity has
  signed off on the URL's digest in Rekor
- `sget sign <URL>` to keylessly sign the URL+digest using a Fulcio OIDC
  flow, with the signature appended to Rekor
- `sget trust` to configure trusted identities
- `sget <OCI>` to fetch blobs in OCI registries (currently without
  checking signatures)

This also adds documentation to the README.

Signed-off-by: Jason Hall <jasonhall@redhat.com>

### CODE REVIEW

File: trust.go
Comment (luhring) [inline trust.go]: super nit: should say "remove", right?

**Review by luhring — APPROVED** [review]
I love this! Agree that we should figure out OCI signing/verification before release, but I don't see why we can't merge this for now. This sets up the patterns that we'll want to have consistent between OCI and HTTP.

**Comment by imjasonh** [comment]
It sounds like there may be some scoping/directional questions about this, but overall I'm not hearing much push-back on landing this change and cleaning all that up before a future release. @dlorenc @lukehinds wdyt?

Aside from that I'd really like another set of eyes on the Rekor interaction specifically. I'm worried what I'm writing there now will be overly limiting if I want to check that the digest matches the requested URL, and sorting that out soon would be nice. Also not blocking on this PR, just a general call for Rekor experts to double-check my work.

**Comment by dlorenc** [comment]
> It sounds like there may be some scoping/directional questions about this, but overall I'm not hearing much push-back on landing this change and cleaning all that up before a future release. @dlorenc @lukehinds wdyt?

My stance has always been that we need a pure download/verifier so the signing functionality here worries me about scope creep. If we need it just for testing or something, then sure. But I thought where we landed originally was that sget would just be for downloading and verifying.

To be clear - I'm fine with lots of signing tools for specific use cases, I just really want there to be a downloader/verifier that's lightweight and doesn't bring in all the complexity of signing with different key types and KMS systems and hardware keys.

**Comment by imjasonh** [comment]
If the addition of `sign` worries you, I'm fine moving it into a separate executable, or hiding the CLI surface for now until we can straighten it out. I mostly just want to merge this PR so folks can try it out and tell me what's wrong with it 🙃 

FWIW, [`sget`'s roadmap](https://github.com/sigstore/sget/blob/main/ROADMAP.md) doesn't currently point toward becoming a Universal Verifier, though I _definitely_ agree that's useful, and it's not like the roadmap is set in stone.

If it does become that, I think we need to decide whether it's also responsible for downloading, or just verifying before handing off to some purpose-built downloader. When it's OCI and URLs it's fairly easy, but if this grows to include Rust crates, PyPi modules, Ruby gems, WASM thingies, etc., the "simple, lightweight verifier" also includes a bunch of artitfact-type-specific download logic, which gets cumbersome.

**Comment by luhring** [comment]
Personally, I kind of like having the `sget sign` command built right in. I agree it adds to the user surface area. But I think it makes for a very natural user experience with `sget`.

The [roadmap](https://github.com/sigstore/sget/blob/main/ROADMAP.md) talks about the ultimate goal of maintainers signing their release artifacts / URLs. We need to make this a very low-friction task for maintainers (somehow...), or no one is going to sign anything, and there will be nothing for users to verify.

As a stepping stone before maintainer signing and policy catches on, the roadmap also talks about crowdsourced signatures.

>sget users can gain assurance about the safety of an artifact by relying on crowd-sourced verification and signatures

Any work pushed onto the "crowd" needs to extra low-friction. If users are participating in the sget "ecosystem" by using it to download artifacts, it's very natural for them to use a signing functionality to publicly announce their trust of an artifact **if they don't** need to reach for a separate tool, IMHO. For example, I might install a pet project from `https://jason.io/install.sh`, see that it works and didn't do anything crazy, and then immediately sign the script bytes.

>doesn't bring in all the complexity of signing with different key types and KMS systems and hardware keys

I could see us completing the entire roadmap without ever bringing in KMSes and hardware keys. I don't know if others share this view. 😃  But I think this is a separate dimension from whether or not we include signing.

If we can't think of a way forward with `sget sign` that we all feel comfortable with, I'd like us to find a different way forward. What are other options? Move this signing into cosign?

And re: Universal Verifier, I agree with @imjasonh's take. I'd love to see us get through the whole roadmap (and work out the kinks) with just OCI and URLs before considering additional ecosystems to plug into.

**Comment by imjasonh** [comment]
I generally agree with everything @luhring is saying (as usual 🙄).

I'd like to get this incarnation of `sget` in, including `sget sign`, and worry about its positioning in the ecosystem once we're a little further along. If it ends up becoming our Universal Verifer, that's great, we can move signing out to a separate tool. But for now it's just sitting in a PR 😄

**Pull request merged**
